Automattic/woocommerce-payments
# Improve: has_multi_currency_orders query improvement and unit tests

Fixes #6814

#### Changes proposed in this Pull Request



Added `Analytics:->has_multi_currency_orders` method query improvement and unit tests.


This PR adds an improvement to the query found in `has_multi_currency_orders` method where instead of doing a count and having to scan all rows, we look for the first occurrence of `_wcpay_multi_currency_order_exchange_rate` meta key and return the result.

The functionality of that method and expected response is not modified, only the underlying query to generate the necessary results.


Nothing is expected to break this code, if results are found, true is returned, otherwise false.

Added a unit test to test functionality.


#### Testing instructions



The query is enabled on all rest API endpoints, example `/wp-json/wc/store/v1/cart/`. 

- Ensure the code changes in Analytics.php have been applied correctly by running the unit tests for the WCPay_Multi_Currency_Analytics_Tests class.
- Verify that the has_multi_currency_orders() method now correctly checks for multi-currency orders in both default and custom order tables.
- Test the analytics functionality on the WooCommerce admin pages and ensure that the Multi-Currency related support is working as expected.
- Check that the customer currencies list is correctly populated in the WooCommerce settings and accessible from the front-end JS script.

There are currently no known issues or side effects related to these code changes. The code has been carefully reviewed and tested to ensure proper functionality and compatibility.

#### Performance comparrison

Original query:

```
#> explain analyze SELECT COUNT(post_id) FROM wp_postmeta WHERE meta_key = '_wcpay_multi_currency_order_exchange_rate';

-> Aggregate: count(wp_postmeta.post_id)  (cost=160055.35 rows=148020) (actual time=251.008..251.008 rows=1 loops=1)
    -> Filter: (wp_postmeta.meta_key = '_wcpay_multi_currency_order_exchange_rate')  (cost=145253.35 rows=148020) (actual time=0.272..247.841 rows=85683 loops=1)
        -> Index lookup on wp_postmeta using meta_key (meta_key='_wcpay_multi_currency_order_exchange_rate')  (cost=145253.35 rows=148020) (actual time=0.271..214.589 rows=85683 loops=1)
```

and the new suggested query

```
#> explain analyze SELECT EXISTS(SELECT 1 FROM wp_postmeta WHERE meta_key = '_wcpay_multi_currency_order_exchange_rate' limit 1) as count ;

-> Rows fetched before execution  (cost=0.00..0.00 rows=1) (actual time=0.000..0.000 rows=1 loops=1)
-> Select #2 (subquery in projection; run only once)
    -> Limit: 1 row(s)  (cost=146023.84 rows=1) (actual time=0.426..0.426 rows=1 loops=1)
        -> Filter: (wp_postmeta.meta_key = '_wcpay_multi_currency_order_exchange_rate')  (cost=146023.84 rows=148800) (actual time=0.425..0.425 rows=1 loops=1)
            -> Index lookup on wp_postmeta using meta_key (meta_key='_wcpay_multi_currency_order_exchange_rate')  (cost=146023.84 rows=148800) (actual time=0.423..0.423 rows=1 loops=1)
```

notice now the first query needs 251.008 miliseconds to aggregate the count and has to process 148020 rows while the proposed query only needs to process 1 row and executes much faster.

Time measurement

```
#> select exists(SELECT 1 FROM wp_postmeta WHERE meta_key = '_wcpay_multi_currency_order_exchange_rate' limit 1 ) as count;
+-------+
| count |
+-------+
|     1 |

#>  SELECT COUNT(post_id) FROM wp_postmeta WHERE meta_key = '_wcpay_multi_currency_order_exchange_rate';
+----------------+
| COUNT(post_id) |
+----------------+
|          85685 |
+----------------+
1 row in set (0.256 sec)
```

The validation of that function (85685 > 0) === (1 === 1)



* details are added in the initial issue #6814 where old query and new query execution and load metrics are  compared.

-------------------


**Post merge**
